src/screencast/screencast.c

```diff
@@ -0,0 +1,241 @@
+#include "screencast.h"
+
+static struct screencast_context ctx;
+
+static const char object_path[] = "/org/freedesktop/portal/desktop";
+static const char interface_name[] = "org.freedesktop.impl.portal.ScreenCast";
+
+#define SDB_CHECK(x)  do { ret = x; if(ret < 0) { printf("Line %d: error %d\n", __LINE__, ret); return ret; }} while(0)
+
+int setup_outputs(struct screencast_context *ctx) {
+
+  int output_id;
+
+  struct wayland_output *output, *tmp_o;
+  wl_list_for_each_reverse_safe(output, tmp_o, &ctx->output_list, link) {
+    printf("Capturable output: %s Model: %s: ID: %i\n", output->make,
+           output->model, output->id);
+    output_id = output->id;
+  }
+
+  output = wlr_find_output(ctx, NULL, output_id);
+  if (!output) {
+    printf("Unable to find output with ID %i!\n", output_id);
+    return 1;
+  }
+
+  ctx->target_output = output;
+  ctx->framerate = output->framerate;
+  ctx->with_cursor = true;
+
+  printf("wl_display fd: %d\n", wl_display_get_fd(ctx->display));
+  
+  return 0;
+
+}
+
+void *start_screencast(void *data){
+
+	struct screencast_context *ctx = data;
+
+  pthread_mutex_init(&ctx->lock, NULL);
+
+  wlr_register_cb(ctx);
+
+  pthread_create(&ctx->pwr_thread, NULL, pwr_start, ctx);
```

Comment:
In pwr_start, I'm pretty sure the call to pw_main_loop_run will tie up execution, so we can't get to our while loop where we call wl_display_dispatch. I'll double check. Basically, we are running our while loop for screencopy, running the pipewire loop, and listening/responding to d-bus messages, so I think we need the original process thread and two more to support those three concurrent operations.

Reply:
It should be possible to integrate pipewire's event loop into our main loop, of the other way around. I'll have a look at the pipewire API.

Reply:
So right now, I trigger pipewire events on frame ready from WLR. Maybe if we switched back to a timer like in the video-src.c example, we could dispatch from there. I'll have to think about this.

Reply:
Removed the pipewire thread. Only one left for dbus now. Because of the volume of changes, this review no longer points at anything meaningful, so I'll mark it resolved, but there's more to fix.